### ISSUE TITLE
Bad response to rangeFormatting/textEdits?

### ISSUE THREAD

**Issue by sebastiansturm** [emacs-lsp/lsp-mode#1138]
I'm seeing occasional clangd crashes/aborts where clangd complains about invalid offsets within some source file. This just happened again and looking at the `lsp-log` turned up the following pattern:

1. I'm editing some part of the source file, say I'm inserting a semicolon or pasting code from the clipboard
2. in response to the change just made, `lsp-mode` emits a `rangeFormatting` request referring to version 45 of the source file (I had previously unsuccessfully tried disabling as many formatting options as I could find, but perhaps a similar issue had been caused by other factors, such as change-hook using minor modes; will reinvestigate once this issue is resolved)
3. clangd responds with a `rangeFormatting` response containing 3 TextEdits
4. `lsp-mode` sends out a `didChange` message, the content and location of which are perfectly correct when taking the previous 3 TextEdits into account (I checked by logging all intermediate buffer versions to a temporary variable); also, the reported version is now at 49.

At this point clangd aborts, I guess due to `lsp-mode` failing to send out the 3 intermediate `onChange` messages (?)
When stepping through `lsp--apply-text-edit-replace-buffer-contents` in a similar situation, `lsp-mode` appears to behave correctly but I noticed that it emits the modified buffer wholesale because `lsp--apply-formatting` temporarily sets `lsp--server-sync-method` to `'full`.

So I wonder:
- why temporarily switching to full syncs is necessary? Are incremental changes somehow unreliable when used for buffer formatting?
- if this, in combination with full sync debouncing, causes a race condition between the 3 textEdit responses and whatever `didChange` message should have been sent out in response to my original editing operation? Or maybe switching back to incremental is supposed to flush pending full syncs and something's wrong with my setup (didn't check the code yet to see if that is the case)?

**Comment by yyoncho** [comment]
The reason for using full sync was a perf bug - formatting in relatively big file was taking 15+ seconds. Based on your description I think that we should flush before preparing incremental updates or right after the formatting. We flush before a call to the server but apparently the changes.(I am over the phone, cannot go into details)

**Comment by sebastiansturm** [comment]
thanks for the quick response! Maybe it would make sense to temporarily switch to full sync for large edits, but stick with incremental sync for small formatting responses (let's say we stick with incremental if the number of textEdits is < 10 or so and their total length is below some reasonable threshold?). If you think that's ok, I can try that change locally and open a PR if it seems to work for me

**Comment by yyoncho** [comment]
This make sense but it is more about optimization. It will avoid the issue but IMO.we still have to fix the root cause.

**Comment by sebastiansturm** [comment]
absolutely, sorry if my writing was unclear. If we wish to change sync methods on the fly, that should of course be done correctly, but if this switching of sync methods happens to strongly affect autoformatting performance (and autoformatting seems to be the only operation that changes syncing behavior on the fly as far as I can tell), combining both changes into a single "fix-autoformatting" PR might be a reasonable thing to do? Though if you'd rather split this into two separate changes (bugfix + performance tweak), that's also fine with me.

As for when to best do the flushing, I'm on the fence too. Checking whether we have to flush on every change message probably only takes microseconds or so (?) and protects us from falling into the same trap the next time a temporary switch to full syncs is needed for some other operation. On the other hand, I don't like the idea of burdening the often-invoked `lsp-on-change` with extra logic (and making it harder to read in the process) whereas `lsp--apply-formatting` is a trivial two-line function that could easily accommodate a few extra lines without degrading in readability (and I guess `lsp--apply-formatting` isn't often called several times in close succession so getting rid of unneeded flushes between subsequent calls would provide little benefit). So I'm slightly in favor of having `lsp--apply-formatting` check whether it needs to flush (which I assume is the case if we're running in incremental mode on function entry) and, if yes, to flush on function exit

**Comment by sebastiansturm** [comment]
would it make sense to do the temporary switch to full syncs every time we invoke `lsp--apply-text-edits` (I guess some refactoring operations such as renaming may also produce large numbers of TextEdits and hence cause performance issues)? Also, the docstring of `lsp--apply-text-edits` says "This method is used if we do not have `buffer-replace-content'." which doesn't seem to be correct. If you're fine with it, I'll remove that sentence when I make the change?

**Comment by yyoncho** [comment]
I think it make sense to have the same logic but let's not do it until we are sure that the issue is fixed. The apply changes code is used on a lot of places and I do not want to introduce regression there. About the comment: I am not familiar with that piece of code and ATM I do not have access to desktop pc and I could not check the code I will get back to you later

**Comment by yyoncho** [comment]
On the previous question: fixing the issue in the apply formatting function seems to be the simplest and safest option. Feel free to organise your commits the way it is more convenient for you. I have noticed a performance issue when doing find replace so long term we might consider delaying incremental updates too. This has been discussed in the past and we decided not to do it because the server has to have denounce either way but we might revisit this decision.

**Comment by sebastiansturm** [comment]
ok, I'll first push a minimal change to my own `lsp-mode` fork without opening a PR since I'll be away until Sunday; will then have a look at how to best set up a regression test

**Comment by sebastiansturm** [comment]
do we have some sort of lsp protocol verifier that we can apply to the io-log generated by `lsp-mode`? Otherwise I could try and add one; in the present issue this should clearly show the error (though we'd still need a reliable and preferably cross-platform way to simulate user input)

**Comment by yyoncho** [comment]
> do we have some sort of lsp protocol verifier that we can apply to the io-log generated by lsp-mode? 

No, we don't. IMO it will be easier to plug by adding proxies on top of the methods for sending/receiving requests/notifications instead of parsing the logs.

**Comment by sebastiansturm** [comment]
I think I got a reproducible test case. If you wish to try it out, please open the following C++ file with a clangd language server (I'm using a trunk build, v10 branch):
```
int some_function() {
   auto x =
   some_long_function_name(123)
}
```
Then execute the following function:
```
(defun break-lsp () (interactive)
       ;; increasing debounce interval so the bug is easier to trigger
       ;; manually if this is desired
       (setq lsp-debounce-full-sync-notifications-interval 30.0
             ;; disable features that cause full-sync flushes via
             ;; lsp--send-request-async
             lsp-eldoc-enable-hover nil
             lsp-enable-symbol-highlighting nil
             lsp-eldoc-enable-signature-help nil)
       (lsp-format-buffer)
       (goto-char 59)
       (insert ";"))
```
Unless your clang-formatting options somehow happen to be completely different from mine (I used clangd's defaults for this test), buffer formatting should cause the `some_long_function_name` to end up at line 1, s.t. the `didChange` event emitted by the inserted semicolon will cause clangd to emit the following error message to stderr (on a typical setup, the `utf-8` will read `utf-16`; I configured clangd to use utf 8 offsets, but that is irrelevant to this issue):
```
Failed to update <...>.cpp: utf-8 offset 58 is invalid for line 0
```
I got a small ruby script that checks the IO log for invalid offsets so a clangd installation isn't needed for the unit test. Could convert that to emacs lisp I guess. Any suggestions on how to best fit the test into `lsp-mode`'s existing testing infrastructure?

**Comment by yyoncho** [comment]
I tried it on my local setup with clangd-8 and I didn't saw the error. I will update the clangd version in the lsp-docker image and will try it there. But if you are consistently reproducing the issue IMO this is not that important.


> Any suggestions on how to best fit the test into lsp-mode's existing testing infrastructure?

Option 1: start cland suite similar to what we have for pyls. The test could be - execute `break-lsp` (it could be easily converted to equkes steps) and then perform an operation that will work only if lsp state is not broken. 

Option 2: write "server" that can apply the changes and verify that the file, in the end, is the correct one. That would require mocking some stuff. IMO this is better than writing tests that verify that the changes result in the correct file than to write test that verifies that lsp-mode sends particular set of changes. 

IMO both approaches would require relatively a lot of work.
